File: handler/player.py
# -*- coding: utf-8 -*-
"""reAudioPlayer ONE"""
__copyright__ = ("Copyright (c) 2022 https://github.com/reAudioPlayer")

import asyncio

import pygame
from aiohttp import web

from dataModel.song import Song
from db.dbManager import DbManager
from player.player import Player
from player.playerPlaylist import PlayerPlaylist
from player.playlistManager import PlaylistManager


class PlayerHandler:
    """player handler"""
    def __init__(self,
                 player: Player,
                 playlistManager: PlaylistManager,
                 dbManager: DbManager) -> None:
        self._player = player
        self._playlistManager = playlistManager
        self._dbManager = dbManager

    async def getPlay(self, _: web.Request) -> web.Response:
        """get(/api/player/play)"""
        asyncio.create_task(self._player.play())
        return web.Response(status = 200, text = "success!")

    async def getPause(self, _: web.Request) -> web.Response:
        """get(/api/player/pause)"""
        asyncio.create_task(self._player.pause())
        return web.Response(status = 200, text = "success!")

    async def getPlayPause(self, _: web.Request) -> web.Response:
        """get(/api/player/playPause)"""
        asyncio.create_task(self._player.playPause())
        return web.Response(status = 200, text = "success!")

    async def getNext(self, _: web.Request) -> web.Response:
        """get(/api/player/next)"""
        asyncio.create_task(self._player.next())
        return web.Response(status = 200, text = "success!")

    async def getLast(self, _: web.Request) -> web.Response:
        """get(/api/player/previous)"""
        asyncio.create_task(self._player.last())
        return web.Response(status = 200, text = "success!")

    async def loadPlaylist(self, request: web.Request) -> web.Response:
        """post(/api/player/load)"""
        x = await request.json()
        if x.get("type") == "playlist":
            asyncio.create_task(self._player.loadPlaylist(self._playlistManager.get(x["id"])))
        elif x.get("type") == "collection":
            asyncio.create_task(self._player.loadPlaylist(PlayerPlaylist.liked(self._dbManager)))
        elif x.get("type") == "collection/breaking":
            asyncio.create_task(self._player.loadPlaylist(PlayerPlaylist.breaking(self._dbManager)))
        elif x.get("type") == "track":
            print(f"id={x['id']}")
            asyncio.create_task(self._player
                .loadPlaylist(PlayerPlaylist(self._dbManager,
                                             songs = self._dbManager\
                                                .getSongsByCustomFilter(f"id={x['id']}"),
                                             name = x['id'])))
        return web.Response(status = 200, text = "success!")

    async def setVolume(self, request: web.Request) -> web.Response:
        """post(/api/player/volume)"""
        x = await request.json()
        pygame.mixer.music.set_volume(int(x["value"]) / 100.0)
        return web.Response(status = 200, text = "success!")

    async def getVolume(self, _: web.Request) -> web.Response:
        """get(/api/player/volume)"""
        return web.Response(status = 200, text = str(round(pygame.mixer.music.get_volume() * 100)))

    async def loadSongAt(self, request: web.Request) -> web.Response:
        """post(/api/player/at)"""
        x = await request.json()
        async def _implement() -> None:
            print(x.get("type"))
            if "playlistIndex" in x:
                if not await self._player.loadPlaylist(
                        self._playlistManager.get(x.get("playlistIndex")),
                                                  x["index"]):
                    await self._player.at(x["index"])
            elif x.get("type") == "collection/breaking":
                print("do breaking!!")
                if not await self._player.loadPlaylist(PlayerPlaylist.breaking(self._dbManager),
                                                       x["index"]):
                    await self._player.at(x["index"])
                print("hello??")
            elif x.get("type") == "collection":
                if not await self._player.loadPlaylist(PlayerPlaylist.liked(self._dbManager),
                                                       x["index"]):
                    await self._player.at(x["index"])
            await self._player.at(x["index"])

        asyncio.create_task(_implement())
        return web.Response(status = 200, text = "success!")

    async def updateSong(self, request: web.Request) -> web.Response:
        """post(/api/tracks/{id})"""
        id_ = int(request.match_info['id'])
        jdata = await request.json()
        self._player.updateSongMetadata(id_, Song.fromDict(jdata))
        return web.Response(status = 200, text = "success!")

    async def setShuffle(self, request: web.Request) -> web.Response:
        """post(/api/player/shuffle)"""
        x = await request.json()
        self._player.shuffle = bool(x["value"])
        return web.Response(status = 200, text = "success!")

    async def getShuffle(self, _: web.Request) -> web.Response:
        """get(/api/player/shuffle)"""
        return web.Response(status = 200, text = str(self._player.shuffle))

    async def setLoopSong(self, request: web.Request) -> web.Response:
        """post(/api/player/repeat)"""
        x = await request.json()
        self._player.loopSong = bool(x["value"])
        return web.Response(status = 200, text = "success!")

    async def getLoopSong(self, _: web.Request) -> web.Response:
        """get(/api/player/repeat)"""
        return web.Response(status = 200, text = str(self._player.loopSong))

    async def setPos(self, request: web.Request) -> web.Response:
        """post(/api/player/seek)"""
        x = await request.json()
        self._player.setPos(x["value"])
        return web.Response(status = 200, text = "success!")

    async def getPos(self, _: web.Request) -> web.Response:
        """get(/api/player/seek)"""
        return web.Response(status = 200, text = str(self._player.getPos()))

File: src/ui/src/store/player.ts
/*
 * Copyright (c) 2023, reAudioPlayer ONE.
 * Licenced under the GNU General Public License v3.0
 */

import { defineStore } from "pinia";
import { parseCover, zeroPad } from "../common";
import { useDataStore } from "./data";
import { SharedPlayer } from "../api/sharedPlayer";
import { getShuffle, nextSong, prevSong, setShuffle } from "../api/player";
import { computed } from "vue";
import { type ILyrics, findLyrics } from "../views/SingAlong/lyrics";


type PlaylistType = "playlist" | "collection" | "collection/breaking" | "track";
export type RepeatType = "repeat" | "repeat_one_on" | "repeat_on";
export interface Playable {
    play: () => void;
    pause: () => void;
    seek: (time: number) => void;
    setVolume: (volume: number) => void;
    setRepeat?: (repeat: RepeatType) => void; // handled in store
    setMute: (mute: boolean) => void;
}


export const usePlayerStore = defineStore({
    id: 'player',
    state: () => ({
        playing: false,
        progress: 0,
        ready: false,
        muted: false,
        shuffle: false,
        song: {
            title: null,
            artist: null,
            album: null,
            cover: parseCover(null),
            source: null,
            duration: null,
            favourite: false,
            id: -1,
            metadata: {
                plays: 0,
                spotify: {
                    id: null,
                }
            }
        },
        playlist: {
            cover: null,
            description: null,
            name: null,
            id: -1,
            songs: [],
        },
        lyrics: null as ILyrics | null,
        volume: 50,
        repeat: "repeat" as RepeatType,
        sharedPlayer: null as InstanceType<typeof SharedPlayer> | null,
        player: null as Playable | null,
    }),
    actions: {
        playPause() {
            if (this.playing) {
                this.pause();
            } else {
                this.play();
            }
        },
        toggleRepeat() {
            switch (this.repeat) {
                case "repeat":
                    this.repeat = "repeat_on";
                    break;
                case "repeat_on":
                    this.repeat = "repeat_one_on";
                    break;
                case "repeat_one_on":
                    this.repeat = "repeat";
                    break;
            }
            localStorage.setItem("reap.repeat", this.repeat);
            this.player?.setRepeat(this.repeat);
        },
        setRepeat(repeat: RepeatType) {
            this.repeat = repeat;
        },
        setMute(mute: boolean) {
            this.muted = mute;
            this.player?.setMute(mute);
        },
        toggleMute() {
            this.setMute(!this.muted);
        },
        setShuffle(shuffle: boolean) {
            this.shuffle = shuffle;
            setShuffle(shuffle);
        },
        toggleShuffle() {
            this.setShuffle(!this.shuffle);
        },
        onSongEnded() {
            if (this.repeat === "repeat_one_on") {
                this.play();
            } else {
                if (this.repeat === "repeat" && this.playlist.index.value === this.playlist.songs.length - 1) {
                    return;
                }

                this.next();
            }
        },
        play() {
            this.player.play();
        },
        pause() {
            this.player.pause();
        },
        next() {
            nextSong();
        },
        previous() {
            prevSong();
        },
        setPlayer(player: Playable) {
            if (!player) return;
            if (this.player === player) return;

            this.player = player;
        },
        async setSong(song) {
            if (song.id == this.song.id) return;
            this.song = song;
            this.song.cover = parseCover(song.cover);
            this.progress = 0;

            this.lyrics = await findLyrics();
        },
        setReady(ready) {
            if (this.ready === ready) return;

            this.ready = ready;

            if (ready) {
                useDataStore().initialise();
            } else {
                this.$reset();
                useDataStore().$reset();
            }
        },
        setDuration(duration) {
            this.song.duration = duration;

            fetch(`/api/tracks/${this.song.id}`, {
                method: "PUT",
                body: JSON.stringify({
                    duration
                })
            })

            //saveDuration(this.song.id, duration);
        },
        setPlaying(playing) {
            this.playing = playing;
        },
        seek(time) {
            this.player.seek(time);
        },
        seekPercent(percent) {
            this.seek(this.durationSeconds * percent / 100);
        },
        setProgress(progress) {
            this.progress = Math.round(progress);
        },
        setFavourite(favourite) {
            this.song.favourite = favourite;
            fetch(`/api/tracks/${this.song.id}`, {
                method: "PUT",
                body: JSON.stringify({
                    favourite
                })
            })
        },
        setPlaylist(playlist) {
            this.playlist.songs = playlist.songs;
            this.playlist.cover = parseCover(playlist.cover);
            this.playlist.description = playlist.description;
            this.playlist.name = playlist.name;
            this.playlist.id = playlist.id;
        },
        setVolume(volume) {
            if (volume == this.volume) return;

            this.volume = volume;
            localStorage.setItem("reap.volume", volume);
            if (this.player.setVolume) {
                this.player?.setVolume(volume);
            }
        },
        toggleFavourite() {
            this.setFavourite(!this.song.favourite);
        },
        async initialise() {
            this.volume = localStorage.getItem("reap.volume") || 50;
            this.repeat = localStorage.getItem("reap.repeat") || "repeat_on";
            this.sharedPlayer = new SharedPlayer();

            this.setShuffle(await getShuffle());
        },
        // TODO
        // loadPlaylist( ??? )
        loadPlaylist(playlistId: number | PlaylistType, id: number = null) {
            const body = {
                type: "playlist",
                id: playlistId,
            }

            if (typeof playlistId === "string") {
                body.type = playlistId;
                body.id = id;
            }

            fetch("/api/player/load", {
                method: "POST",
                body: JSON.stringify(body)
            })
        },
        loadSong(playlist: number | PlaylistType, index: number) {
            const body = {
                index
            }

            if (typeof playlist === "number") {
                if (!isNaN(playlist)) {
                    body["playlistIndex"] = playlist;
                }
            } else {
                body["type"] = playlist;
            }

            fetch("/api/player/at", {
                method: "POST",
                body: JSON.stringify(body)
            });
        }
    },
    getters: {
        hasLyrics(state) {
            return !state.lyrics?.error;
        },
        durationSeconds(state) {
            return state.song.duration;
        },
        displayDuration(state) {
            const duration = state.song.duration;
            if (isNaN(duration)) return "0:00";
            return `${Math.floor(duration / 60)}:${zeroPad(Math.floor(duration % 60), 2)}`
        },
        stream(state) {
            return `/api/player/stream/${state.song.id}`;
        },
        cover(state) {
            return state.song.cover;
        },
        progressPercent(state) {
            return state.progress / this.durationSeconds * 1000;
        },
        displayProgress(state) {
            const progress = state.progress;
            if (isNaN(progress)) return "0:00";
            return `${Math.floor(progress / 60)}:${zeroPad(Math.floor(progress % 60), 2)}`
        },
        loaded(state) {
            return state.song.id != -1;
        },
        shuffleIcon(state) {
            if (state.shuffle) {
                return "shuffle_on";
            }
            return "shuffle";
        },
        muteIcon(state) {
            if (state.muted) {
                return "volume_off";
            }
            if (state.volume > 50) {
                return "volume_up";
            }
            if (state.volume > 0) {
                return "volume_down";
            }
            return "volume_mute";
        },
        playlist(state) {
            return {
                ...state.playlist,
                index: computed(() => state.playlist?.songs?.findIndex(song => song.id === state.song.id) ?? -1)
            };
        }
    }
});
